core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1417,21 +1426,35 @@ class KafkaApis(val requestChannel: RequestChannel,
     val transactionalId = addPartitionsToTxnRequest.transactionalId
     val partitionsToAdd = addPartitionsToTxnRequest.partitions
 
-    // Send response callback
-    def sendResponseCallback(error: Errors): Unit = {
-      def createResponse(throttleTimeMs: Int): AbstractResponse = {
-        val responseBody: AddPartitionsToTxnResponse = new AddPartitionsToTxnResponse(throttleTimeMs, error)
-        trace(s"Completed $transactionalId's AddPartitionsToTxnRequest with partitions $partitionsToAdd: errors: $error from client ${request.header.clientId}")
-        responseBody
+    if(!authorize(request.session, Write, new Resource(ProducerTransactionalId, transactionalId)))
+      sendResponseMaybeThrottle(request, (throttleTimeMs: Int) => new InitPidResponse(throttleTimeMs, Errors.TRANSACTIONAL_ID_AUTHORIZATION_FAILED))
+    else {
+      val (_, unAuthorized) = partitionsToAdd.asScala.partition {
+        tp => authorize(request.session, Write, new Resource(Topic, tp.topic))
```

Comment:
Might also be worth forbidding the users to use this API with any of the internal topics.